src/absorb.rs

```diff
@@ -0,0 +1,364 @@
+    /// Return the list as `Vec`
+    fn to_sponge_field_elements_as_vec<F: PrimeField>(&self) -> Vec<F> {
+        let mut result = Vec::new();
+        self.to_sponge_field_elements(&mut result);
+        result
+    }
+
+    /// Specifies the conversion into a list of bytes for a batch. Append the list to `dest`.
+    fn batch_to_sponge_bytes(batch: &[Self], dest: &mut Vec<u8>)
+    where
+        Self: Sized,
+    {
+        for absorbable in batch {
+            absorbable.to_sponge_bytes(dest)
+        }
+    }
+
+    /// Specifies the conversion into a list of bytes for a batch. Return the list as `Vec`.
+    fn batch_to_sponge_bytes_as_vec(batch: &[Self]) -> Vec<u8>
+    where
+        Self: Sized,
+    {
+        let mut result = Vec::new();
+        Self::batch_to_sponge_bytes(batch, &mut result);
+        result
+    }
+
+    /// Specifies the conversion into a list of field elements for a batch. Append the list to `dest`.
+    fn batch_to_sponge_field_elements<F: PrimeField>(batch: &[Self], dest: &mut Vec<F>)
+    where
+        Self: Sized,
+    {
+        for absorbable in batch {
+            absorbable.to_sponge_field_elements(dest)
+        }
+    }
+
+    /// Specifies the conversion into a list of field elements for a batch. Append the list to `dest`.
+    fn batch_to_sponge_field_elements_as_vec<F: PrimeField>(batch: &[Self]) -> Vec<F>
+    where
+        Self: Sized,
+    {
+        let mut result = Vec::new();
+        for absorbable in batch {
+            absorbable.to_sponge_field_elements(&mut result)
+        }
+        result
+    }
+}
+
+/// An extension to `Absorb` that is specific to item with variable length, such as a list.
+pub trait AbsorbWithLength: Absorb {
+    /// The length of the `self` being absorbed.
+    fn absorb_length(&self) -> usize;
+
+    /// Converts the object into a list of bytes along with its length information
+    /// that can be absorbed by a `CryptographicSponge`.
+    /// Append the list to `dest`.
+    fn to_sponge_bytes_with_length(&self, dest: &mut Vec<u8>) {
+        self.absorb_length().to_sponge_bytes(dest);
+        self.to_sponge_bytes(dest)
+    }
+
+    /// Converts the object into field elements along with its length information
+    /// that can be absorbed by a `CryptographicSponge`.
+    /// Append the list to `dest`
+    fn to_sponge_field_elements_with_length<F: PrimeField>(&self, dest: &mut Vec<F>) {
+        self.absorb_length().to_sponge_field_elements(dest);
+        <Self as Absorb>::to_sponge_field_elements(&self, dest)
+    }
+}
+
+/// If `F1` equals to `F2`, return `x` as `F2`, otherwise panics.
+/// ## Panics
+/// This function will panic if `F1` is not equal to `F2`.
+pub(crate) fn field_cast<F1: PrimeField, F2: PrimeField>(input: F1) -> F2 {
+    if TypeId::of::<F1>() != TypeId::of::<F2>() {
+        panic!("Trying to absorb non-native field elements.")
+    } else {
+        F2::from_le_bytes_mod_order(&to_bytes!(input).unwrap())
+    }
+}
```

Comment:
```suggestion
/// If `F1` and `F2` are the same type, return `input` but cast to `F2`, and panic otherwise.
/// ## Panics
/// This function will panic if `F1` is not equal to `F2`.
pub(crate) fn field_cast<F1: PrimeField, F2: PrimeField>(input: F1) -> F2 {
    if TypeId::of::<F1>() != TypeId::of::<F2>() {
        panic!("Trying to absorb non-native field elements.")
    } else {
        F2::from_le_bytes_mod_order(&to_bytes!(input).unwrap())
    }
}
```

Question: should we change this to return an `Option`? That is, return `None` in the case that the fields are unequal. Also, should we change the condition to be something more general like `if F1::characteristic() != F2::characteristic()`? It doesn't matter too much since it's an internal API, maybe we can do it after accepting the PR.

Reply:
@Pratyush returning `None` makes sense to me (because in some cases the caller might know if the field type is the same). I will change to that

Reply:
I've fixed this. Yes, I think using `characteristic` to check field equality works because current API only supports prime field